pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator/index.jelly

```diff
@@ -0,0 +1,102 @@
+                    for top-level directives, or into a <code>stage</code> block for stage directives. See
+                    <a href="https://jenkins.io/doc/pipeline/syntax/" target="_blank">the online syntax documentation</a>
+                    for more information on directives.
+                </f:block>
+                <f:section title="${%Directives}"/>
+                <!-- Similar to f:dropdownDescriptorSelector, but adds fallback content to block, and JENKINS-25130 adds per-selection help: -->
+                <j:set var="item" value="${it.getItem(request)}"/>
+                <d:taglib uri="local">
+                    <d:tag name="listDirectives">
+                        <j:set var="inStage" value="false"/>
+                        <j:forEach var="descriptor" items="${directives}">
+                            <f:dropdownListBlock title="${descriptor.name}: ${descriptor.displayName}" staplerClass="${descriptor.clazz.name}" lazy="descriptor,item,inStage">
+                                <l:ajax>
+                                    <j:set var="it" value="${item}"/>
+                                    <j:set var="instance" value="${null}"/>
+                                    <j:set var="help" value="${descriptor.helpFile}"/>
+                                    <j:set var="inStage" value="false"/>
+                                    <j:if test="${help != null}">
+                                        <tr>
+                                            <td colspan="3"/>
+                                            <f:helpLink url="${help}"/>
+                                        </tr>
+                                        <f:helpArea/>
+                                    </j:if>
+                                    <st:include from="${descriptor}" page="${descriptor.configPage}" optional="true">
+                                        <f:block>
+                                            ${%This directive has not yet defined any visual configuration.}
+                                        </f:block>
+                                    </st:include>
+                                </l:ajax>
+                            </f:dropdownListBlock>
+                        </j:forEach>
+                    </d:tag>
+                </d:taglib>
+                <f:dropdownList name="prototype" title="${%Sample Directive}">
+                    <local:listDirectives directives="${it.getDirectives()}"/>
+                </f:dropdownList>
+                <j:set var="id" value="${h.generateId()}"/>
+                <f:block>
+                    <input type="button" value="${%Generate Declarative Directive}" onclick="handlePrototype_${id}(); return false" class="submit-button primary"/>
+                    <f:textarea id="prototypeText_${id}" readonly="true" style="margin-top: 10px" />
+                    <script>
+                        function handlePrototype_${id}() {
```

Comment:
would be nice if thid was in it's own `.js` file so it can be linted during build. Maybe overkill.

Reply:
also, whitespaces please :)

Reply:
Yeah, this is all copy-pasted/tweaked from Snippetizer's stuff, so I didn't make any more changes than I had to in case that resulted in the world exploding.